### ISSUE TITLE
Re-factored qualified name extraction logic for golang to make it consistent with the java analogous lib. [phab:D2165]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2627]
_Created by **craigbarber** at 2018-02-02 23:20:08_:

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2018-02-02 23:22:51_:

I think you might want `delim` instead of `ctx.PostChildText`.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2018-02-02 23:32:18_:

Please add a test for an empty `PostChildText` if you meant `delim` here.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2018-02-02 23:32:18_

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-02-02 23:34:46_:

It's counter-intuitive, but using PostChildText here is intentional. delim is only used for the package name join. Whereas if no postChildText is specified, (as is the case for process nodes), we don't want to use a default delim. This is consistent with the Java lib as well.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-02-02 23:36:43_:

There are two existing test cases with empty PostChildTexts.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-02-03 01:12:41_:

```
if got := RenderQualifiedName(&ms); !proto.Equal(got, test.want) {
```

It's not safe to compare protos by value (though in this case it will probably work most of the time) because the generated code includes details that affect the form but not the meaning. 

You don't need to check for nil. It's fine to compare to a nil of a message type, it behaves as an empty message of the type.

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-02-03 01:12:41_:

I'm still not fond of using a proto for this, but now that we've already defined it, I guess we might as well use it consistently.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-02-05 17:48:02_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-02-06 05:29:06_:

Dead now?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-02-06 05:29:06_:

Since you always want a pointer anyway:

    info := &cpb.SymbolInfo{BaseName: id.PreText}

then the below can simply be

    return info

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-02-06 05:29:06_:

Another option you can consider is to always return a non-nil value, even if it's otherwise empty:

    return new(cpb.SymbolInfo)

They're equivalent as proto values (i.e., they report equal under `proto.Equal`) but always returning a value can sometimes make it easier for the caller who doesn't have to check if they want to override the value. But it's up to you.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-02-06 17:41:10_:

incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-02-06 17:43:00_:

What's the recommended approach for testing equality of protos?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-02-06 17:48:18_:

What you have here works fine: `proto.Equal`

The https://godoc.org/github.com/google/go-cmp/cmp package also affords some support for diffing, though in this case the messages are small enough it's simpler to just print them.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2018-02-06 17:48:18_

**Comment by kythe** [comment]
_**craigbarber** closed via commit 4bcd3a3b947b75e2f09a81ecd6b79e903fcd4cd4 at 2018-02-06 20:42:06_
